modules/concurrency/include/hephaestus/concurrency/queue_consumer.h

```diff
@@ -0,0 +1,69 @@
+//=================================================================================================
+// Copyright (C) 2023-2024 HEPHAESTUS Contributors
+//=================================================================================================
+
+#pragma once
+
+#include <functional>
+#include <thread>
+
+#include "hephaestus/containers/blocking_queue.h"
+
+namespace heph::concurrency {
+
+/// QueueConsumer creates a thread that wait for messages to be pushed in the queue and call the user
+/// provided callback on them.
+template <typename T>
+class QueueConsumer {
+public:
+  using Callback = std::function<void(const T&)>;
+  QueueConsumer(Callback&& callback, std::optional<std::size_t> max_queue_size);
```

Comment:
```suggestion
  [[nodiscard]] QueueConsumer(Callback&& callback, std::optional<std::size_t> max_queue_size);
```
perhaps it makes sense to mark the constructor nodiscard, since the thread  is joined when going class goes out of scope?

Reply:
Interesting, I never realized you could mark a constructor nodiscard.
Alternativly we could use the start/stop pattern.